src/cursor/cur_hs.c

```diff
@@ -10,30 +10,61 @@
 #include "wt_internal.h"
 
 /*
- * __wt_hs_cursor_open --
- *     Open a new history store table cursor.
+ * __hs_cursor_open_int --
+ *     Open a new history store table cursor, internal function.
  */
-int
-__wt_hs_cursor_open(WT_SESSION_IMPL *session)
+static int
+__hs_cursor_open_int(WT_SESSION_IMPL *session, WT_CURSOR **cursorp)
 {
     WT_CURSOR *cursor;
     WT_DECL_RET;
     const char *open_cursor_cfg[] = {WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL};
 
-    /* Not allowed to open a cursor if you already have one */
-    WT_ASSERT(session, session->hs_cursor == NULL);
-
     WT_WITHOUT_DHANDLE(
       session, ret = __wt_open_cursor(session, WT_HS_URI, NULL, open_cursor_cfg, &cursor));
     WT_RET(ret);
 
     /* History store cursors should always ignore tombstones. */
     F_SET(cursor, WT_CURSTD_IGNORE_TOMBSTONE);
 
-    session->hs_cursor = cursor;
+    *cursorp = cursor;
+    return (0);
+}
+
+/*
+ * __wt_hs_cursor_cache --
+ *     Cache a new history store table cursor. Open and then close a history store cursor without
+ *     saving it in the session.
+ */
+int
+__wt_hs_cursor_cache(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *cursor;
+
+    conn = S2C(session);
+    if (F_ISSET(conn, WT_CONN_IN_MEMORY) || F_ISSET(session, WT_SESSION_NO_RECONCILE) ||
+      (session->dhandle != NULL && WT_IS_METADATA(S2BT(session)->dhandle)) ||
```

Comment:
Just curious:  Why the check for a metadata file dhandle?

Reply:
The original conditional also checked and skipped for the metadata, I just moved it here. The metadata is often handled specially. However, I will pull more/most of the original comment about the conditions this conditional. That comment did not mention the metadata but gives other context.

Reply:
Most likely because we do not have history store content for metadata, so we do not need to cache history store dhandle if we are evicting metadata.